core/modules/contextual/js/contextual.js

```diff
@@ -16,30 +16,107 @@ Backdrop.behaviors.contextualLinks = {
+      });
+
+      // Reset margins on all wrappers.
+      allWrappers.forEach(function(info) {
+        info.wrapper.css('margin', '0');
+      });
+
+      // Recalculate margins to avoid collisions.
+      var dir = $('html').attr('dir');
+      const hsize = 28; // width of trigger wrapper
+      const vsize = 19; // height of trigger wrapper
+      var n = allWrappers.length;
+      for (let i = 0; i < n; i++) {
+        var follower = allWrappers[i];
+        // Compare follower against all of its predecessors in the list (any of
+        // which may have already been adjusted).
+        for (let j = 0; j < i; j++) {
+          var leader = allWrappers[j];
+          // Adjust the position of follower if necessary to avoid collision
+          // with leader.
+          var leaderOffset = leader.wrapper.offset();
+          var followerOffset = follower.wrapper.offset();
+          var verticalOverlap = followerOffset.top >= leaderOffset.top && followerOffset.top < leaderOffset.top + vsize;
+          if (dir == 'ltr') {
+            var horizontalOverlap = followerOffset.left >= leaderOffset.left - hsize && followerOffset.left < leaderOffset.left + hsize;
+            if (verticalOverlap && horizontalOverlap) {
+              // We have a collision; shift the follower down if there's room,
+              // otherwise left.
+              if (followerOffset.top + 2 * vsize <= follower.regionOffsetBottom) {
+                // Shift down
+                follower.vshift += vsize;
+                follower.wrapper.css('margin-top', follower.vshift);
+              }
+              else {
+                // Shift left and start a new column.
+                follower.vshift = 0;
+                follower.hshift += hsize;
+                follower.wrapper.css('margin-top', follower.vshift);
+                follower.wrapper.css('margin-right', follower.hshift);
+              }
+            }
+          }
+          else { // rtl
+            var horizontalOverlap = followerOffset.left > leaderOffset.left - hsize && followerOffset.left <= leaderOffset.left + hsize;
+            if (verticalOverlap && horizontalOverlap) {
+              // We have a collision; shift the follower down if there's room,
+              // otherwise right.
+              if (followerOffset.top + 2 * vsize <= follower.regionOffsetBottom) {
+                // Shift down
+                follower.vshift += vsize;
+                follower.wrapper.css('margin-top', follower.vshift);
+              }
+              else {
+                // Shift right and start a new column.
+                follower.vshift = 0;
+                follower.hshift += hsize;
+                follower.wrapper.css('margin-top', follower.vshift);
+                follower.wrapper.css('margin-left', follower.hshift);
+              }
+            }
+          }
+        }
+      }
+    }
+    $(document).ready(adjustContextualLinks);
+    $(window).on('resize', adjustContextualLinks);
```

Comment:
We have a method for preventing the resize handler from firing repeatably on resize:

```suggestion
    Backdrop.optimizedResize.add(adjustContextualLinks, 'adjustContextualLinks');
```